SwiftCompilerSources/Sources/Optimizer/FunctionPasses/AutodiffClosureSpecialization.swift

```diff
@@ -0,0 +1,716 @@
+    // ClosureSpecializer and CapturePropagation.
+    // CapturePropagation propagates constant function-literals. Such
+    // function specializations can then be optimized again by the
+    // ClosureSpecializer and so on.
+    // This happens if a closure argument is called _and_ referenced in
+    // another closure, which is passed to a recursive call. E.g.
+    //
+    // func foo(_ c: @escaping () -> ()) {
+    //   c()
+    //   foo({ c() })
+    // }
+    //
+    // A limit of 2 is good enough and will not be exceed in "regular"
+    // optimization scenarios.
+    let closureCallee =
+      closure is PartialApplyInst
+      ? (closure as! PartialApplyInst).referencedFunction!
+      : (closure as! ThinToThickFunctionInst).referencedFunction!
+
+    if closureCallee.specializationLevel > specializationLevelLimit {
+      continue
+    }
+
+    if callSiteMap[pai] == nil {
+      callSiteMap.insert(key: pai, value: CallSite(applySite: pai))
+    }
+
+    intermediateClosureArgDescriptors.append(
+      IntermediateClosureArgDescriptor(
+        callSite: pai,
+        closureArgumentIndex: closureArgumentIndex,
+        parameterInfo: closureParamInfo,
+        reachableExitBBs: Array(callee.blocks.filter { $0.isReachableExitBlock })
+      )
+    )
+  }
+
+  finalizeCallSites(
+    context,
+    for: closure,
+    callSiteMap: &callSiteMap,
+    lifeRangeEndpoints: lifeRangeEndpoints,
+    intermediateClosureArgDescs: intermediateClosureArgDescriptors
+  )
+
+  return false
+}
+
+/// Handles conversion and reabstraction uses of the supported closure.
+///
+/// - Parameters:
+///   - context: `FunctionPassContext` instance
+///   - for: A closure against which callees may be specialized.
+/// - Returns:
+///   - remainingUses: Any uses not handled in this function.
+///   - lifeRangeEndpoints: Instructions that comprise the possible life-range of the closure.
+///   - possibleMarkDependenceBases: The possible mark_dependence bases of the closure.
+///   - haveUsedReabstraction: Whether the closure has gone throug a reabstraction.
+private func handleNonApplyUses(
+  _ context: FunctionPassContext,
+  for closure: SingleValueInstruction
+) -> (
+  remainingUses: [Operand], lifeRangeEndpoints: [Instruction],
+  possibleMarkDependenceBases: ValueSet, haveUsedReabstraction: Bool
+) {
+  var initialUses = Array(closure.uses)
+  var remainingUses: [Operand] = []
+  var lifeRangeEndpoints: [Instruction] = []
+  var haveUsedReabstraction = false
+
+  // Set of possible arguments for mark_dependence. The base of a
+  // mark_dependence we copy must be available in the specialized function.
+  var possibleMarkDependenceBases = ValueSet(context)
+
+  if let pai = closure as? PartialApplyInst {
+    for arg in pai.arguments {
+      possibleMarkDependenceBases.insert(arg)
+    }
+  }
+
+  // Uses may grow in this loop to handle transitive uses
+  for useIndex in 0..<initialUses.count {
```

Comment:
It should be possible to use a `ValueWorklist` or `InstructionWorklist` for this purpose. It would also avoid that uses are visited twice.

Reply:
I need operands here because I need to access both the value being used and the owning instruction. And it looks like it's not possible to have an `IntrusiveSet` and therefore a `WorkList` for operands?

Reply:
You can use a `ValueWorklist` and append the values for which you want to see the uses (basically all values, where you now do a `initialUses.append(contentsOf: value.uses)`). E.g.
```
  var worklist = ValueWorklist(context)
  worklist.pushIfNotVisited(closure)
  while let value = worklist.pop() {
    for use in value.uses {
      ...
      worklist.pushIfNotVisited(cfi)
      ...
      worklist.pushIfNotVisited(cvt)
      ...
    }
  }
```

Reply:
While this does help us with memory allocations, I'm not fully sure how this helps us avoid visiting uses twice?

We will still go over the subset of uses we "reject" here in the `handleApplies` function, no?